lib/EmitVerilog/EmitVerilog.cpp

```diff
@@ -852,6 +852,12 @@ class ExprEmitter : public ExprVisitor<ExprEmitter, SubExprInfo>,
   SubExprInfo visitComb(rtl::ShlOp op) {
     return emitSignedBinary(op, Shift, "<<");
   }
+  SubExprInfo visitComb(rtl::ShrsOp op) {
+    return emitSignedBinary(op, Shift, ">>");
+  }
+  SubExprInfo visitComb(rtl::ShruOp op) {
+    return emitSignedBinary(op, Shift, ">>>");
```

Comment:
I didn't notice this before, but is `emitSignedBinary` the right function here?  My understanding is that verilog will always give the right answer regardless of the signedness of the operands.  

Please check out the difference between `emitSignedBinary` and `emitBinary`.  It would also be good to write an IR testcase that illustrates the difference and make sure the right Verilog is generated.

Checking things like `(shl (x divs y), (x divs y))` and `(shru (x divs y), (x divu y))` and other permutations would be good to make sure that the verilog printer is properly modeling the semantics of the IR.

Reply:
I don't think I'm following here, which is why I've made this mistake. 

1. I see that other uses have it simply as ```emitBinary``` for ```>>``` and ```<<```. However, I don't understand why ```hasStrictSign``` isn't always true. We wouldn't want to do something like ```1 >> -1``` (this is UB in C-family languages if I recall correctly). Why does sign not matter in, say, a (variadic) multiplication, but it does in a division?

2.  I'm unsure what you mean by ```x divs y```. So you want to do something like ```%a div %b```, where ```a, b``` have a signed type, say, ```si1```? And similar for ```x divu y```, except we use, say, ```iu1```?

Reply:
I am not certain this is a mistake - I don't know verilog well enough.

Here's the issue :). Verilog has a notion of whether something is signed or not.  For example, the result of `$signed(foo)` is considered to be signed.

Some operations perform the same computation regardless of whether or not their operands are signed or unsigned.  I suspect that `>>>` and `>>` and `<<` are this way.  If so, then it is always safe to emit `x << y` because the signs of x and y don't matter.

However, some operations change their behavior if the operands are signed or unsigned.  I think that the comparisons are like this: `$signed(x) > $signed(y)` is different when x/y are unsigend than a simple comparison.

In the verilog emitter, there are two important things to get right: 1) for an operation like comparison, we need to make sure the operands are emitted with the expected sign.   2) for every operation, the returned SubExprInfo needs to correctly indicate whether the result was signed or not.

This difference is how `emitSignedBinary/emitBinary` work.  emitSignedBinary (which is really just emitBinary with hasStrictSign=true) is based on the *FIRRTL* type system notion of sign: SInt vs UInt.  However, the RTL dialect doesn't have any of this.  We need to derive the sign of the operation from the operation itself (its opcode), which means that we may need another "emitBinary" equivalent for operations (like comparisons) that care about the operand sign, and we need a way to set the result sign that doesn't depend on the firrtl type either.

Reply:
FWIW, the Vivado HLS solution to this is that almost all variables are described as bit vectors, and signed operations explicitly make their inputs $signed().  In our experience this was reasonably readable and understandable, even if it might not be the particular preference of every Verilog designer out there.

Reply:
@lattner Got it, that makes a bit more sense now. The differences between signed representation for FIRRTL and RTL was not something I was aware of.